### PR TITLE
WT-10157 Aggregate record count and byte count for a table

### PR DESCRIPTION
**Pull request by GWlodarek**

### CODE REVIEW

File: src/session/session_api.c [hunk 1]
```diff
@@ -833,16 +833,42 @@ __session_alter_readonly(WT_SESSION *wt_session, const char *uri, const char *co
...
+        if ((ret = __wt_meta_checkpoint(session, uri, NULL, &ckpt)) != 0)
+            WT_ERR(ret == WT_NOTFOUND ? ENOENT : ret);
+
+        *count = ckpt.ps.row_count;
+    } else if (WT_PREFIX_SKIP(uri, "table:")) {
+        if ((ret = __wt_schema_get_table(session, uri, strlen(uri), false, 0, &table)) != 0)
+            WT_ERR(ret == WT_NOTFOUND ? ENOENT : ret);
+
+        /* Set the number of rows from one column group, sum the column group byte counts. */
+        for (i = 0; i < WT_COLGROUPS(table); i++) {
```
Comment (quchenhao) [inline src/session/session_api.c:858]: Why we need to sum the for the column group? Should we actually support this api for complex tables?

File: src/session/session_api.c [hunk 2]
```diff
@@ -833,16 +833,42 @@ __session_alter_readonly(WT_SESSION *wt_session, const char *uri, const char *co
 int
 __wt_session_count(WT_SESSION *wt_session, const char *uri, int64_t *count)
```
Comment (quchenhao) [inline src/session/session_api.c:834]: This api only supports row count not byte count?

File: src/session/session_api.c [hunk 3]
```diff
@@ -833,16 +833,42 @@ __session_alter_readonly(WT_SESSION *wt_session, const char *uri, const char *co
...
+    if (WT_PREFIX_MATCH(uri, "file:")) {
+        if ((ret = __wt_meta_checkpoint(session, uri, NULL, &ckpt)) != 0)
+            WT_ERR(ret == WT_NOTFOUND ? ENOENT : ret);
+
+        *count = ckpt.ps.row_count;
+    } else if (WT_PREFIX_SKIP(uri, "table:")) {
+        if ((ret = __wt_schema_get_table(session, uri, strlen(uri), false, 0, &table)) != 0)
+            WT_ERR(ret == WT_NOTFOUND ? ENOENT : ret);
+
+        /* Set the number of rows from one column group, sum the column group byte counts. */
```
Comment (quchenhao) [inline src/session/session_api.c:857]: If we only support row count, why we aggregate the bytes counts? I think this is not implemented.

File: test/suite/test_count01.py [hunk 4]
```diff
@@ -31,16 +31,55 @@
...
+from wtscenario import make_scenarios
 
 class test_count01(wttest.WiredTigerTestCase):
     tablename = 'test_count01'
-    uri = 'table:' + tablename
+    tablename2 = 'test'
+
+    keyfmt = [
+        ('integer-row', dict(keyfmt='i')),
+        ('column', dict(keyfmt='r')),
```
Comment (quchenhao) [inline test/suite/test_count01.py:43]: Should we increase test for more types of tables? Fixed length column store and column groups if we support it.

File: src/btree/bt_slvg.c [hunk 5]
```diff
@@ -676,7 +680,15 @@ __slvg_trk_leaf(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, uint8_t *ad
     case WT_PAGE_ROW_LEAF:
         WT_TIME_AGGREGATE_INIT_MERGE(&trk->trk_ta);
         WT_CELL_FOREACH_KV (session, dsk, unpack) {
-            WT_TIME_AGGREGATE_UPDATE(session, &trk->trk_ta, &unpack.tw);
+            switch (unpack.type) {
```
Comment (quchenhao) [inline src/btree/bt_slvg.c:683]: Should we do anything here? I think we will automatically aggregate when we reconcile the page.

File: src/include/btmem.h [hunk 6]
```diff
@@ -134,6 +134,13 @@ struct __wt_page_stat {
     /* These informational values can be negative to signify that they are invalid. */
     int64_t byte_count;
     int64_t row_count;
+
+    /*
+     * A page's page stat information may need to be reset if we find that any of its children do
+     * not have valid page stat values.
+     */
+    bool reset_byte_count;
```
Comment (quchenhao) [inline src/include/btmem.h:142]: Can we use a flag variable for this?

File: src/reconcile/rec_col.c [hunk 7]
```diff
@@ -56,19 +56,24 @@ __wt_bulk_insert_fix(WT_SESSION_IMPL *session, WT_CURSOR_BULK *cbulk, bool delet
...
     cursor = &cbulk->cbt.iface;
+    WT_PAGE_STAT_INIT(&ps);
 
     WT_RET(__rec_col_fix_bulk_insert_split_check(cbulk));
     __bit_setv(
       r->first_free, cbulk->entry, btree->bitcnt, deleted ? 0 : ((uint8_t *)cursor->value.data)[0]);
     ++cbulk->entry;
     ++r->recno;
 
+    ps.row_count = 1;
```
Comment (quchenhao) [inline src/reconcile/rec_col.c:74]: I don't understand why we need to set it to 1 first.

Comment (quchenhao) [reply src/reconcile/rec_col.c:74]: I think we only need to aggregate when we try to write out the leaf page.

Comment (quchenhao) [reply src/reconcile/rec_col.c:74]: Same for row store and column store in the bulk load case.

File: src/reconcile/rec_row.c [hunk 8]
```diff
@@ -606,6 +627,9 @@ __rec_row_leaf_insert(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins)
         }
         WT_TIME_AGGREGATE_UPDATE(session, &r->cur_ptr->ta, &tw);
 
+        ps.row_count = 1;
```
Comment (quchenhao) [inline src/reconcile/rec_row.c:630]: I also don't understand why this is needed.

File: src/session/session_api.c (hunk 3 above)
Comment (mm-ng) [reply src/session/session_api.c:857]: Yes, currently the byte count part is not yet implemented.

File: src/session/session_api.c (hunk 2 above)
Comment (mm-ng) [reply src/session/session_api.c:834]: Yes, that's correct. I need to clarify with Vamsi whether it's better to add this as a different API or into the same one.

File: src/reconcile/rec_col.c (hunk 7 above)
Comment (mm-ng) [reply src/reconcile/rec_col.c:74]: I'm setting it to 1 to increment the row count when we call update. I was just trying to follow the places where we update the time aggregate as well, but we can probably discuss this more.

File: src/session/session_api.c (hunk 1 above)
Comment (mm-ng) [reply src/session/session_api.c:858]: I'm not sure whether this should be supported or not, I'll get back to this later.

File: src/reconcile/rec_col.c [hunk 9]
```diff
@@ -323,9 +342,16 @@ __wt_rec_col_int(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *pageref)
         if (page_del != NULL)
             WT_TIME_AGGREGATE_MERGE(session, &r->cur_ptr->ta, &ft_ta);
         WT_TIME_AGGREGATE_MERGE(session, &r->cur_ptr->ta, &ta);
+        WT_PAGE_STAT_UPDATE(&r->cur_ptr->ps, &ps);
     }
     WT_INTL_FOREACH_END;
 
+    /* Reset the page stat value for the page if any of its children had invalid stat values. */
```
Comment (kommiharibabu) [inline src/reconcile/rec_col.c:349]: Resetting here can be wrong. We already have split the page above. Instead of storing the flag and resetting it later,  reset there itself inside the `WT_PAGE_STAT_UPDATE` macro and do not update further if the reset flag is set.

Comment (mm-ng) [reply src/reconcile/rec_col.c:349]: Ok, thanks for the suggestion.

File: src/btree/bt_slvg.c (hunk 5 above)
Comment (mm-ng) [reply src/btree/bt_slvg.c:681]: I think this was already discussed offline that we need it.

File: test/suite/test_count01.py (hunk 4 above)
Comment (mm-ng) [reply test/suite/test_count01.py]: I have added a fixed-length column store scenario. I'm not sure if we can test the column groups scenario, given that it won't be accurate all the time.

File: src/session/session_api.c (hunk 1 above)
Comment (quchenhao) [reply src/session/session_api.c:858]: We can test that. There are python tests test colgroup.

File: src/reconcile/rec_col.c [hunk 10]
```diff
@@ -323,6 +339,8 @@ __wt_rec_col_int(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *pageref)
         if (page_del != NULL)
             WT_TIME_AGGREGATE_MERGE(session, &r->cur_ptr->ta, &ft_ta);
         WT_TIME_AGGREGATE_MERGE(session, &r->cur_ptr->ta, &ta);
+        if (__wt_process.page_stats_2022)
```
Comment (kommiharibabu) [inline src/reconcile/rec_col.c:342]: why check the flag here? Based on the flag we do not write to disk already. Is there any specific reason behind it?

File: src/meta/meta_ckpt.c [hunk 11]
```diff
@@ -1143,13 +1153,14 @@ __wt_meta_ckptlist_to_meta(WT_SESSION_IMPL *session, WT_CKPT *ckptbase, WT_ITEM
...
-          ",prepare=%d,write_gen=%" PRId64 ",run_write_gen=%" PRId64 ")",
+          ",prepare=%d,write_gen=%" PRId64 ",run_write_gen=%" PRId64 ",byte_count=%" PRId64
+          ",row_count=%" PRId64 ")",
           (int)ckpt->addr.size, (char *)ckpt->addr.data, ckpt->order, ckpt->sec,
           (int64_t)ckpt->size, (int64_t)ckpt->ta.newest_start_durable_ts,
           (int64_t)ckpt->ta.oldest_start_ts, (int64_t)ckpt->ta.newest_txn,
           (int64_t)ckpt->ta.newest_stop_durable_ts, (int64_t)ckpt->ta.newest_stop_ts,
           (int64_t)ckpt->ta.newest_stop_txn, (int)ckpt->ta.prepare, (int64_t)ckpt->write_gen,
-          (int64_t)ckpt->run_write_gen));
+          (int64_t)ckpt->run_write_gen, ckpt->ps.byte_count, ckpt->ps.row_count));
```
Comment (kommiharibabu) [inline src/meta/meta_ckpt.c:1163]: Do we want to control the checkpoint data update also with a feature flag?

File: src/reconcile/rec_write.c [hunk 12]
```diff
@@ -2500,7 +2507,10 @@ __rec_write_wrapup(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_PAGE *page)
          */
         ref = r->ref;
         if (__wt_ref_is_root(ref)) {
-            __wt_checkpoint_tree_reconcile_update(session, &ta);
+            if (__wt_process.page_stats_2022)
```
Comment (kommiharibabu) [inline src/reconcile/rec_write.c:2510]: Can you add a comment to describe why we are setting the count to zero

File: src/meta/meta_ckpt.c (hunk 11 above)
Comment (mm-ng) [reply src/meta/meta_ckpt.c:1163]: Yes, I believe that's preferable. I don't think we have access to the flags stored in the page header here but maybe I can use the `__wt_process.page_stats_2022` flag instead.

File: src/reconcile/rec_write.c (hunk 12 above)
Comment (mm-ng) [reply src/reconcile/rec_write.c:2510]: I thought the checkpoint information was not persisting correctly in the scenario of an empty table, but seems like it might be because we're not initialising empty tables properly in the reconciliation flow. I'll need to look at this again to confirm.

File: src/meta/meta_ckpt.c (hunk 11 above)
Comment (kommiharibabu) [reply src/meta/meta_ckpt.c:1163]: Yes, the use of `__wt_process.page_stats_2022` is enough. When we enable this flag, it starts writing on the page and also to the checkpoint.

**Comment by mm-ng** [comment]
Given that I’ll be on leave soon, here’s a summary of where the PR is at @bvpvamsikrishna 
- I believe the current implementation is working correctly for tables that contain records inside them, but the row count is not being saved correctly when tables are empty. I haven’t been able to find the correct place to initialise this as it seems to get overridden when I try to initialise it before reconciling an internal page, even when there are no records in the table. It seems like the code is traversing through some other pages which get merged into the internal page and invalidates the row count of the internal page.
- There is a Python test that tests the count API implemented. I don’t intend on merging this test in as we do not have the feature flag enabled in the develop code, it is currently just being used to test the API during development.
- The count API generates the correct count result when inserting records into a table, though it may be worth testing more edge cases.
- It may also be good to test column groups if we intend to support this, I haven’t gotten around to that yet.

File: src/reconcile/rec_write.c (hunk 12 above)
Comment (mm-ng) [reply src/reconcile/rec_write.c:2510]: I think the checkpoint information is different to the information that is packed onto the disk. Here I think we should set to zero since we are checkpointing an empty page. Currently, it seems that the checkpoint information is correct though I have doubts the zero is being packed to disk correctly.

File: src/reconcile/rec_col.c [hunk 13]
```diff
@@ -159,6 +167,9 @@ __wt_bulk_insert_var(WT_SESSION_IMPL *session, WT_CURSOR_BULK *cbulk, bool delet
     /* Initialize the time aggregate that's going into the parent page. See note above. */
     WT_TIME_AGGREGATE_UPDATE(session, &r->cur_ptr->ta, &tw);
 
+    ps.row_count = (int64_t)cbulk->rle;
+    WT_PAGE_STAT_UPDATE(&r->cur_ptr->ps, &ps);
```
Comment (quchenhao) [inline src/reconcile/rec_col.c:171]: Here the row_count is the delta to add. WT_PAGE_STAT_UPDATE is a confusing name. May be rename it to WT_PAGE_STAT_ADD

File: src/include/pagestat_inline.h [hunk 14]
```diff
@@ -23,3 +23,61 @@
...
+            (ps)->row_count = WT_STAT_NONE;                                          \
+        else                                                                         \
+            (ps)->row_count += (merge)->row_count;                                   \
+    } while (0)
+
+/*
+ * Update the page stat structure. For both the byte and row counts, we want to reset the page stat
+ * value for the page if any of its children that were merged in had invalid stat values.
+ */
+#define WT_PAGE_STAT_UPDATE(ps, update)                       \
```
Comment (quchenhao) [inline src/include/pagestat_inline.h:46]: Is there a case we pass update with -1 count?

File: src/include/pagestat_inline.h [hunk 15]
```diff
@@ -23,3 +23,61 @@
 
 /* Copy the values from one time page stat structure to another. */
 #define WT_PAGE_STAT_COPY(dest, source) (*(dest) = *(source))
+
+/* Merge a page stat structure into another - summing the value from each. */
+#define WT_PAGE_STAT_MERGE(ps, merge)                                                \
+    do {                                                                             \
+        /* Merge byte count */                                                       \
+        if ((ps)->byte_count == WT_STAT_NONE || (merge)->byte_count == WT_STAT_NONE) \
+            (ps)->byte_count = WT_STAT_NONE;                                         \
```
Comment (quchenhao) [inline src/include/pagestat_inline.h:32]: Why we don't need to set the flag WT_RESET_ROW_COUNT

File: src/reconcile/rec_col.c [hunk 16]
```diff
@@ -323,6 +339,8 @@ __wt_rec_col_int(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *pageref)
         if (page_del != NULL)
             WT_TIME_AGGREGATE_MERGE(session, &r->cur_ptr->ta, &ft_ta);
         WT_TIME_AGGREGATE_MERGE(session, &r->cur_ptr->ta, &ta);
+        if (__wt_process.page_stats_2022)
+            WT_PAGE_STAT_UPDATE(&r->cur_ptr->ps, &ps);
```
Comment (quchenhao) [inline src/reconcile/rec_col.c:343]: Sometimes we use merge, sometimes we use update. They are really just two functions with mostly similar functionality with a little difference. Can we combine them?

File: src/reconcile/rec_col.c [hunk 17]
```diff
@@ -941,6 +961,9 @@ __wt_rec_col_fix(
              * No need to have a minimum split size boundary, all pages are filled 100% except the
              * last, allowing it to grow in the future.
              */
+            ps.row_count = entry;
+            WT_PAGE_STAT_UPDATE(&r->cur_ptr->ps, &ps);
+
```
Comment (quchenhao) [inline src/reconcile/rec_col.c:966]: I don't understand why we cannot use merge here.

File: src/btree/bt_slvg.c (hunk 5 above)
Comment (quchenhao) [reply src/btree/bt_slvg.c:681]: This is more complicated than I thought. By checking the code, we need to handle the size of overflow items and they may not be available in the corrupted data. I think we can ignore aggregating in salvage and let letter checkpoint to aggregate the count for us.

File: src/btree/bt_slvg.c [hunk 18]
```diff
@@ -1225,6 +1234,8 @@ __slvg_col_build_internal(WT_SESSION_IMPL *session, uint32_t leaf_cnt, WT_STUFF
 
         WT_ERR(__wt_calloc_one(session, &addr));
         WT_TIME_AGGREGATE_COPY(&addr->ta, &trk->trk_ta);
+        addr->ps.byte_count = trk->trk_size;
```
Comment (quchenhao) [inline src/btree/bt_slvg.c:1237]: It's wrong to use trk_size as the byte count.

File: src/reconcile/rec_col.c (hunk 16 above)
Comment (mm-ng) [reply src/reconcile/rec_col.c:352]: Yes, I think you're correct we can probably merge the two macros into one.

File: src/include/pagestat_inline.h (hunk 15 above)
Comment (mm-ng) [reply src/include/pagestat_inline.h:32]: We should probably set that flag but I'll combine the macros instead.

File: src/reconcile/rec_col.c (hunk 13 above)
Comment (mm-ng) [reply src/reconcile/rec_col.c]: Yes, I can rename this macro if it makes things easier.

File: src/include/pagestat_inline.h (hunk 14 above)
Comment (mm-ng) [reply src/include/pagestat_inline.h:46]: I think it is possible. When we initialise a `ps` structure it will have -1 values, and it may not get updated to a valid count before calling the macro.

File: src/btree/bt_slvg.c [hunk 19]
```diff
@@ -676,7 +709,37 @@ __slvg_trk_leaf(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, uint8_t *ad
     case WT_PAGE_ROW_LEAF:
         WT_TIME_AGGREGATE_INIT_MERGE(&trk->trk_ta);
         WT_CELL_FOREACH_KV (session, dsk, unpack) {
-            WT_TIME_AGGREGATE_UPDATE(session, &trk->trk_ta, &unpack.tw);
+            switch (unpack.type) {
+            case WT_CELL_KEY:
+                key_size = (int64_t)unpack.size;
+                break;
+            case WT_CELL_KEY_OVFL:
+                key_size = unpack.ovfl_ps.byte_count;
```
Comment (quchenhao) [inline src/btree/bt_slvg.c:717]: Here I think I did it wrong. We don't write byte count for key overflow items. To make this work, we have to write the byte count for key overflow items as well.

Comment (quchenhao) [reply src/btree/bt_slvg.c:717]: I looked at the code again and I don't think we need to aggregate the counts in salvage. Basically, what salvage does is to find any usable leaf page and connects that to a new internal page to build a new btree. If the leaf page has the aggregation already, we don't need to aggregate that again. If it doesn't, we don't need to aggregate that either, just still leave it as invalid until next time it is reconciled. There is also a code path in salvage that `__slvg_row_build_leaf ` merges keys that are usable from multiple leaf pages. In this code path, we do a reconciliation that should automatically aggregate the counts for us. For the internal pages, they are completely rebuilt in memory and will be reconciled later. Therefore, I think there is no need for us to handle aggregation in salvage.

@mm-ng @bvpvamsikrishna

Comment (quchenhao) [reply src/btree/bt_slvg.c:717]: I think I would prefer to write the byte count also for overflow keys. Since overflow key should be rare, it should not generate much extra disk usage. The simplification of the logic is much valuable I think.
